lib/datadog/core/remote/component.rb

```diff
@@ -22,26 +22,80 @@ def initialize(settings, agent_settings)
 
           capabilities = Client::Capabilities.new(settings)
 
+          @barrier = Barrier.new
+
           @client = Client.new(transport_v7, capabilities)
-          @worker = Worker.new(interval: settings.remote.poll_interval_seconds) { @client.sync }
+          @worker = Worker.new(interval: settings.remote.poll_interval_seconds) do
+            begin
+              @client.sync
+            rescue StandardError => e
+              Datadog.logger.error do
+                "remote worker error: #{e.class.name} #{e.message} location: #{Array(e.backtrace).first}"
+              end
+
+              # client state is unknown, state might be corrupted
+              @client = Client.new(transport_v7, capabilities)
+
+              # TODO: bail out if too many errors?
+            end
+
+            @barrier.lift
```

Comment:
Just thinking out loud. Should the lift be just right after `@client.sync`? 


Imagine this scenario. In the rack middleware, we want to wait for the first sync using `Datadog::Core::Remote.active_remote.barrier(:once)`

Imagine we hit `@client.sync` and something fails. We are going to rescue the exception and call `@barrier.lift` 
We are going to continue without having sync against remote configuration. 

I can understand we don't want that because if the error is not transient and is not recovered, we would hold that request indefinitely. 

Just want to leave this train of thought here

Reply:
> I can understand we don't want that because if the error is not transient and is not recovered, we would hold that request indefinitely.

Exactly, that's what we absolutely do _not_ want.

It's a best effort to ensure that in most cases at least one sync call has happened. Even when sync goes through the backend reply may or may not include the full expected configuration, but we _can't_ know if what we receive is what the user expects (e.g it might be legitimately empty or it might be empty because it's simply delayed by rate limiting).